### PR TITLE
Feat/import tables

### PR DESCRIPTION
**Pull request by ianlcz**

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -70,6 +72,38 @@ def report(ctx, output):
     exporter.generate(project, 'report.html')
     pass
 
+@cli.command()
+@click.option('--version-file', default='version.csv', help='CSV Version file')
+@click.option('--overwrite', default=False, help='Overwrite database')
+@click.pass_context
+def import_csv(ctx, version_file, overwrite):
```
Comment (bbalet) [inline main.py]: This should be nammed "import"

File: main.py [hunk 2]
```diff
@@ -70,6 +72,38 @@ def report(ctx, output):
     exporter.generate(project, 'report.html')
     pass
 
+@cli.command()
```
Comment (bbalet) [inline main.py:75]: You should have an additionnal option such as target-table

File: main.py [hunk 3]
```diff
@@ -70,6 +72,38 @@ def report(ctx, output):
...
+@click.option('--version-file', default='version.csv', help='CSV Version file')
+@click.option('--overwrite', default=False, help='Overwrite database')
+@click.pass_context
+def import_csv(ctx, version_file, overwrite):
+    """Import CSV file into tables"""
+    # Overwrite option
```
Comment (bbalet) [inline main.py:82]: All this stuff should be moved into a dedicated package "importers" and you should create a dedicated object. Too long for main and not consistent with the rest of the code.

File: main.py [hunk 4]
```diff
@@ -1,9 +1,11 @@
+from datetime import datetime
 import os
 import sys
 import logging
 import platform
 import subprocess
 import tempfile
+import csv
```
Comment (bbalet) [inline main.py:8]: This is not consistent with the other parts of the code. For instance you've used pandas for importing the JPeek tool output. Why using csv now?

File: main.py [hunk 5]
```diff
@@ -70,6 +72,38 @@ def report(ctx, output):
...
+@click.pass_context
+def import_csv(ctx, version_file, overwrite):
+    """Import CSV file into tables"""
+    # Overwrite option
+    if overwrite:
+        session.query(Version).delete()
+        click.echo("Clearning version table")
+
+    # Read Version CSV file
+    versions = list(csv.DictReader(open(version_file, 'r')))
```
Comment (bbalet) [inline main.py:90]: Check if the file actually exists

File: main.py [hunk 6]
```diff
@@ -70,6 +72,38 @@ def report(ctx, output):
...
+    
+    for version in versions:
```
Comment (bbalet) [inline main.py:94]: What about mandatory fields. I think a pre-check should be done.

File: main.py [hunk 7]
```diff
@@ -70,6 +72,38 @@ def report(ctx, output):
...
+        tag=version["tag"], 
+        start_date=datetime.strptime(version["start_date"], '%Y-%m-%d %H:%M:%S.%f'), 
+        end_date=datetime.strptime(version["end_date"], '%Y-%m-%d %H:%M:%S.%f'), 
+        bugs=version["bugs"], 
+        changes=version["changes"], 
+        avg_team_xp=version["avg_team_xp"], 
+        bug_velocity=version["bug_velocity"])
+
+        # Save on database
+        session.add(newVersion)
```
Comment (bbalet) [inline main.py:107]: You should catch the exception and display a meaningful message for the end user. Especially if a key constraint is violated.

File: main.py (hunk 1 above)
Comment (ianlcz) [reply main.py]: The word _import_ is reserved for python

Comment (bbalet) [reply main.py]: I'm talking about the name of the command

```
@click.command(name='import')
def import_file():
    pass
```

File: importers/flatfile.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    def import_from_csv(self) -> None:
+        """
+        Import CSV to the database
+        """
+        logging.info('import_from_csv')
+        if self.target_table and str(self.target_table).lower() in ["alias", "author", "commit", "file", "issue", 
+                                                      "metric", "model", "ownership", "project", "version"]:
+            # Overwrite option
+            if self.overwrite:
+                self.session.execute('DELETE FROM {target_table}'.format(target_table=str(self.target_table).capitalize()))
```
Comment (bbalet) [inline importers/flatfile.py:34]: Do not put raw SQL query into the code. Use the ORM. Example of delete here: https://github.com/optittm/bugprediction/blob/branches-fusion/connectors/git.py#L180

File: importers/flatfile.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+                click.echo('Overwrite ' + str(self.target_table).capitalize() + ' table')
+
+            if self.filename:
+                # Read CSV file
+                csv_data = list(csv.DictReader(open(self.filename, 'r')))
+                
+                if str(self.target_table).capitalize() == "Version":
+                    for version in csv_data:
+                        save_model(self.session,
```
Comment (bbalet) [inline importers/flatfile.py:43]: You must check if the minimal mandatory fields are here. Especially for version table in which we need:

- start date
- end date
- tag

File: importers/flatfile.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+                if str(self.target_table).capitalize() == "Version":
+                    for version in csv_data:
+                        save_model(self.session, 
+                                    Version(
+                                        project_id=version['project_id'],
+                                        name=version["name"], 
+                                        tag=version["tag"], 
+                                        start_date=datetime.strptime(version["start_date"], '%Y-%m-%d %H:%M:%S.%f'), 
+                                        end_date=datetime.strptime(version["end_date"], '%Y-%m-%d %H:%M:%S.%f'), 
+                                        bugs=version["bugs"],
```
Comment (bbalet) [inline importers/flatfile.py:50]: bugs, changes, xp team and velocity are not imported, these are metrics computed by our tool.

File: importers/flatfile.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+                        save_model(self.session,
+                                    Issue(
+                                        project_id=issue['project_id'],
+                                        number=issue["number"],
+                                        title=issue["title"],
+                                        created_at=datetime.strptime(issue["created_at"], '%Y-%m-%d %H:%M:%S.%f'),
+                                        updated_at=datetime.strptime(issue["updated_at"], '%Y-%m-%d %H:%M:%S.%f')
+                                    ))
+                    click.echo('Importing ' + str(len(csv_data)) + ' issue(s) on database')
+
```
Comment (bbalet) [inline importers/flatfile.py:68]: After the import of either issue or version tables. Metrics related to the version should computed again with `connectors.git.compute_version_metrics()`

File: importers/flatfile.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        logging.info('import_from_csv')
+        if self.target_table and str(self.target_table).lower() in ["alias", "author", "commit", "file", "issue", 
+                                                      "metric", "model", "ownership", "project", "version"]:
+            # Overwrite option
+            if self.overwrite:
+                self.session.execute('DELETE FROM {target_table}'.format(target_table=str(self.target_table).capitalize()))
+                click.echo('Overwrite ' + str(self.target_table).capitalize() + ' table')
+
+            if self.filename:
+                # Read CSV file
```
Comment (bbalet) [inline importers/flatfile.py:38]: in a first loop you should validate all the fields one by one. Are they mandatory ? Is the type correct ? And not wait for an obscure exception to be thrown back at the face of the user.

File: utils/savemodel.py [hunk 13]
```diff
@@ -0,0 +1,8 @@
+import logging
```
Comment (bbalet) [inline utils/savemodel.py:1]: I don't understand why you created a new file in utils for one line of code. Plus the session.commit is into another place.

**Pull request merged**